service/tcp.go

```diff
@@ -285,19 +286,27 @@ func (s *tcpService) handleConnection(listenerPort int, clientTCPConn *net.TCPCo
 			tgtConn.CloseWrite()
 			fromClientErrCh <- fromClientErr
 		}()
-		_, fromTargetErr := ssw.ReadFrom(tgtConn)
-		// Send FIN to client.
-		clientConn.CloseWrite()
-		tgtConn.CloseRead()
 
-		fromClientErr := <-fromClientErrCh
-		if fromClientErr != nil {
+		go func() {
+			_, fromTargetErr := ssw.ReadFrom(tgtConn)
+			// Send FIN to client.
+			clientConn.CloseWrite()
+			tgtConn.CloseRead()
+			fromTargetErrCh <- fromTargetErr
+		}()
+
+		select {
```

Comment:
Oh, it seems like your change has introduced a race condition that is making the tests fail:
<img width="935" alt="image" src="https://user-images.githubusercontent.com/113565/206374310-2de40e64-830b-4949-a676-a181cbb7486a.png">

We need to address that.

Reply:
This is how you can run the tests with the detection of race condition enabled:
https://github.com/Jigsaw-Code/outline-ss-server/blob/4881f8b65f4cee6040f5d07dd412ef7ed29e0a6f/.github/workflows/go.yml#L39